expression/builtin_other.go

```diff
@@ -164,6 +211,152 @@ func (b *builtinCastSig) eval(row []types.Datum) (d types.Datum, err error) {
+	val, isNull, err := b.args[0].EvalInt(row, b.getCtx().GetSessionVars().StmtCtx)
+	return types.NewDecFromInt(val), isNull, errors.Trace(err)
+}
+
+type builtinCastIntAsStringSig struct {
+	baseStringBuiltinFunc
+}
+
+func (b *builtinCastIntAsStringSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalInt(row, b.getCtx().GetSessionVars().StmtCtx)
+	return strconv.FormatInt(val, 10), isNull, errors.Trace(err)
+}
+
+type builtinCastRealAsIntSig struct {
+	baseIntBuiltinFunc
+}
+
+func (b *builtinCastRealAsIntSig) evalInt(row []types.Datum) (res int64, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalReal(row, b.getCtx().GetSessionVars().StmtCtx)
+	return int64(val), isNull, errors.Trace(err)
+}
+
+type builtinCastRealAsDecimalSig struct {
+	baseDecimalBuiltinFunc
+}
+
+func (b *builtinCastRealAsDecimalSig) evalDecimal(row []types.Datum) (res *types.MyDecimal, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalReal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return nil, isNull, errors.Trace(err)
+	}
+	res = new(types.MyDecimal)
+	err = res.FromFloat64(val)
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastRealAsStringSig struct {
+	baseStringBuiltinFunc
+}
+
+func (b *builtinCastRealAsStringSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalReal(row, b.getCtx().GetSessionVars().StmtCtx)
+	return strconv.FormatFloat(val, 'f', -1, 64), isNull, errors.Trace(err)
+}
+
+type builtinCastDecimalAsIntSig struct {
+	baseIntBuiltinFunc
+}
+
+func (b *builtinCastDecimalAsIntSig) evalInt(row []types.Datum) (res int64, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	res, err = val.ToInt()
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastDecimalAsStringSig struct {
+	baseStringBuiltinFunc
+}
+
+func (b *builtinCastDecimalAsStringSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
+	}
+	return string(val.ToString()), false, nil
+}
+
+type builtinCastDecimalAsRealSig struct {
+	baseRealBuiltinFunc
+}
+
+func (b *builtinCastDecimalAsRealSig) evalReal(row []types.Datum) (res float64, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	res, err = val.ToFloat64()
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastStringAsIntSig struct {
+	baseIntBuiltinFunc
+}
+
+func (b *builtinCastStringAsIntSig) evalInt(row []types.Datum) (res int64, isNull bool, err error) {
+	if types.IsHybridType(b.args[0].GetType().Tp) {
```

Comment:
I think we don't need to check this at runtime. If the args0 is hybrid, we needn't add this cast at all, do we ?